Exercise triangle

pclausen opened the pull request:
Adding exercise triangle

kotp commented on config.json:
```diff
@@ -218,20 +218,21 @@
+          "interfaces"
         ]
-      } 
+      }      
```
Why the additional whitespace here?

pclausen replied:
I actually do not know where that whitespace came from. Removed it.

kotp commented on .github/workflows/ci.yml:
```diff
@@ -53,17 +53,19 @@ jobs:
       run: |
-        brew list cmake || brew cask install cmake
-        brew list gfortran || brew cask install gfortran
+        echo "Trying to find latest gfortran version in /usr/local/Cellar/gcc"
```
This is very specific. Is `which` or `whence` appropriate to find out if `gfortran` is available as a command instead?  I mean, do we really care if it is in these areas, or is it OK to be elsewhere, such as `/usr/bin/gfortran`?

pclausen replied:
I am surprised by the Mac system but I think I tried `/usr/bin/gfortran` which did not work. The current way was the best I could think of and that actually works. 

Do you know if it is possible to "debug" the build process in github actions ? I agree that there may be better ways to find the gfortran compiler but "which gfortran" did not work, so this was the second best solution. See also https://github.com/exercism/fortran/pull/126

hiljusti replied:
Can these CI/CD changes be removed from the pull request after #126 ?

kotp replied:
I don't know… I am currently in the process of learning more about github actions myself.  But I just saved this link so that if I come across something, I can get back to you.

kotp replied:
I am talking about locating that, or verifying that it is in the path, though.  `which gfortran` should give a positive answer if it is found in the path.  This is more important I think than the specific location.  Or less brittle should the installation location change.

pclausen replied:
>Can these CI/CD changes be removed from the pull request after #126 ?
Probably but I am too bad at git to make that happen. I would rather suggest to leave the CI/CD stuff for now and clean that up later, if really necessary. It works now and I think it is really hard to debug the CI-stuff.

kotp replied:
I was really hoping to hear some good tips on debugging the CI stuff.

hiljusti replied:
I'll create an issue to revisit & simplify CI/CD

I've seen it getting somewhat major changes across many pull requests and I can't tell what are the same changes or slightly different ones or if they need to get resolved in a certain order etc...

But I think the problem is just that the PRs aren't trying to focus on it. The PRs are interested in adding exercises (and they should remain focused on that) We can just say "Let's come back to it later" and give the CI/CD some proper focus

kotp reviewed (changes requested):
Trailing white space, should be changed, the other is just a question of complexity.

hiljusti commented on config.json:
```diff
@@ -218,20 +218,21 @@
+        "slug": "triangle",
+        "name": "Triangle",
+        "uuid": "58e138e9-55d8-49fa-b05d-70a6aa627b21",
+        "difficulty": 1,
         "practices": [],
         "prerequisites": [],
         "topics" : [
-          "logicals",
+          "bools",
           "conditionals",
           "comparisons",
           "numbers",
-          "bitwise_operations"
+          "sequences",
+          "interfaces"
         ]
-      } 
+      }
```
This might be a merge resolution mistake -- this is overwriting information for the `allergies` exercise instead of adding triangle before or after.

pclausen replied:
I think I resolved this my own merge

hiljusti commented on exercises/practice/triangle/triangle.f90:
```diff
@@ -0,0 +1,12 @@
+
+  contains
+
+  logical function equilateral(tria)
+    integer,dimension(3) :: tria
```
We can afford to use full names, it's 2021 and we don't really need to save bytes in source code.

Instead of `tria` what do you think about `triangle` or `points`?

kotp replied:
@hiljusti is right, but it isn't just about memory, it is about being able to look up words in a dictionary if it is not understood.  If geometry or trigonometry is not my history, then I might think that "tria" is something that "math people" know.  I might try to look it up in my dictionary, but would get little help.  Then I would pick up the phone and ask the author of that line of code, at 3am, because I need to know.  We can all get more sleep if this is a real word.

pclausen replied:
I agree but I took the names from the problem specification which then probably should also be updated.

hiljusti replied:
I think [the problem specification](https://github.com/exercism/problem-specifications/tree/main/exercises/triangle) only defines the function names (`equilateral`, which I'm fine with) and not the parameter names (`tria`, which I think can be `triangle`)

pclausen replied:
I cannot use the word "triangle" because that is the module name, this is the reason or tria. I could maybe change it to "triangle_in" or "triangle_edges". I think I will use the latter.

hiljusti commented on exercises/practice/triangle/triangle.f90:
```diff
@@ -0,0 +1,12 @@
+
+module triangle
```
We should give stubs for all three functions under test. `equilateral`, `isosceles` and `scalene`

hiljusti commented on exercises/practice/triangle/.meta/example.f90:
```diff
@@ -0,0 +1,81 @@
+    intol = abs(a-b)<TOL
+  end function
+
+  logical function istrianle_real(tria)
```
minor: there's a typo. Also I'd suggest maybe `is_triangle_real`

pclausen replied:
fixed

hiljusti commented on exercises/practice/triangle/.docs/instructions.md:
```diff
@@ -0,0 +1,23 @@
+
+A _scalene_ triangle has all sides of different lengths.
+
+## Note
```
We should add a note that this is going to test _both_ for integers and floating point numbers. That's an important detail for what I think this exercise is trying to teach (based on your example)

pclausen replied:
Agreed. Added this note:

> Also, the triangle may be defined with integers or with real values which the
> program should also take into account.

pclausen commented:
@kotp @hiljusti Thank you for your review. I added and modified the changes to the best of my ability but not sure if I addressed all issues. Please check. :-)

hiljusti commented on exercises/practice/triangle/triangle.f90:
```diff
@@ -0,0 +1,53 @@
+
+ contains
+
+  logical function equilateral_real(traingle_edges)
```
Thanks for adding these stubs!

"triangle" is misspelled here. (Also in several places in the example, but I care more about the student-facing material)

Also, thanks for letting me know the variable can't be named "triangle" - I didn't know in Fortran that local variables can't shadow a module name. Makes sense

What do you think about just using the word `edges` or `sides` for the function parameter? I think the name `triangle_edges` is repeating information that should be clear from context

pclausen replied:
duh, thanks. Edit was too quick, sorry about that . "edges" is a much better name.

hiljusti replied:
Not a "duh" moment imo - lots of little changes can make concerns like this less obvious (esp around naming)

kotp commented on exercises/practice/triangle/.meta/example.f90:
```diff
@@ -25,56 +25,56 @@ logical function intol(a,b)
   end function
 
-  logical function istrianle_real(tria)
+  logical function is_trianle_real(traingle_edges)
```
Missing a `g` in the function name, and transposed vowels.

```suggestion
  logical function is_triangle_real(triangle_edges)
```

pclausen replied:
fixed in 54e5349

pclausen commented:
@kotp Thanks, fixed is_trianle_real --> is_triangle_real, but got it wrong in the commit message :-(

pclausen commented:
Not quite sure why the merge is still blocked.... (?)
